### PR TITLE
NSG Script Refactor

### LINKED ISSUE #70: [Task] Refactor NSG Script
Refactor NSG script:
- Remove parameters.json file and replace with hash table input.
- Simplify script and remove user input (region resource checker) aside from script parameters
~~Add logging if there is enough time~~

### PR DESCRIPTION
**Pull request by jleeboyd**
### Description
Resolves #70 - Script refactor
Refactor of PR #45 - Original NSG script

**This PR introduces a script that can be run to create a Network Security Group in Azure using hash tables instead of a separate parameters file.**

### Added features:
- Added in @unicornsleuth login.ps1 script
- Converted parameters.json to hashtable for NSG
- Removed region resource checker function
- 2/3 Added in basic error handling for the two az cmdlets. Also converted to Larissa's login module. Fixed a couple bugs and added in option for custom template path.

### To Test:
**This script requires powershell, to install check out:**
- Windows [Powershell Install](https://docs.microsoft.com/en-us/powershell/scripting/install/installing-powershell?view=powershell-7.1 ) 
- Mac [Homebrew](https://brew.sh/)
1. While in Powershell, in the directory of the script /Automation/nsg, **enter either**:
a.` $./deploy-nsg.ps1` For manual parameter entry. You will then proceed to enter the parameter values (**details for each switch are commented in the deploy-nsg.ps1 param list**).
OR
b. `$./deploy-nsg.ps1 -TenantId  -SPApplicationId -SPSecret -SubscriptionId  -ResourceGroupName  -Location -DeploymentName -OwnerEmail -SecurityGroupName -SecurityRuleName -Protocol -SourcePortRange -DestinationPortRange -SourceAddressPrefix -DestinationAddressPrefix -Access -Priority -Direction` For a quicker deploy, adding in the proper values after the switches (**details for each switch are commented in the deploy-nsg.ps1 param list**).
2. After you have finished with either a. or b. you should now see a successful deploy, given the parameter fields are all correct.
3. Check that your resource group and NSG were properly created in Azure.

**Testing Error handling**
4. Rerun the same command from step 1.a or b. using the same values. You should now see a message telling you to pick a different nsg name.
5. Change your nsg name and rerun the command, you should now see a successful deployment to your rg.
6. You can try different ways to break the deployment (ex. incorrect locations, missing values...), but the errors should be caught and displayed appropriately.
7. **Last but not least be sure to delete the resources created by this script after you are done!**

Time Spent
|**DATE**|**ACTIVITY**|**TIME**|
|---|:---|:---|
|1/25|Initial research into powershell hashtables and defining requirements|2 hours|
|1/26|Continue research and investigate hashtables with ARM templates, start refactoring script|4 hours|
|1/27|Final tests, cleaning up the script+template|2.5 hour|
|2/3|Adapt script to comments|2 hour|

Total time: ~8.5hr + 2hr

### CODE REVIEW

**Review by unicornsleuth — APPROVED** [review]
Everything works! I'll approve this, but please follow up in Slack about variable names, and maybe change those before merging so you don't have to make a separate PR. Also can you change the folder name from NSG to nsg?

File: Automation/NSG/deploy-nsg.ps1 [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+)
+
+$params = @{
+    tags = @{
+        "ownerEmail" = $OwnerEmail
+    }
+    networkSecurityGroups_name = @{
+        "securityGroupName" = $SecurityGroupName
+    }
+    securityRules = @{
```
[the remark this answers is not in the document]
Comment (jleeboyd) [reply Automation/NSG/deploy-nsg.ps1:110]: For now only one rule. I need to look into how to allow for multiple rules dynamically with ARM templates and hash tables. Either this sprint, or next possibly.

File: Automation/NSG/deploy-nsg.ps1 [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+        "destinationPortRanges"= "";     
+        "sourceAddressPrefixes"= "";    
+        "destinationAddressPrefixes"= "";
+#>
+}
+# Logs in and sets subscription      
+& "../login.ps1" $TenantId $ServicePrincipalUsername $ServicePrincipalSecret $SubscriptionId
+
+# Creates/Updates resource group
+New-AzResourceGroup -Name $ResourceGroupName -Location $Location -Force
```
[the remark this answers is not in the document]
Comment (jleeboyd) [reply Automation/NSG/deploy-nsg.ps1:146]: I believe this cmdlet will update the existing resource group, unless this is not the behavior we want?

File: Automation/NSG/template.json [hunk 3]
```diff
@@ -0,0 +1,63 @@
+{
+    "$schema": "https://schema.management.azure.com/schemas/2019-04-01/deploymentTemplate.json#",
+    "contentVersion": "1.0.0.0",
+    "parameters": {
+        "networkSecurityGroups_name": {
+            "type": "object"
```
[the remark this answers is not in the document]
Comment (jleeboyd) [reply Automation/NSG/template.json:6]: No idea, it's a string now!

**Review by zxkevinxz — APPROVED** [review]
Looks like you made all the proper changes.

File: Automation/NSG/deploy-nsg.ps1 [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+    [string]
+    $DestinationPortRanges,
+
+    [Parameter(Mandatory=$False)]
+    [string]
+    $SourceAddressPrefixes,
+
+    [Parameter(Mandatory=$False)]
+    [string]
+    $DestinationAddressPrefies
```
Comment (unicornsleuth) [inline Automation/NSG/deploy-nsg.ps1:106]: Prefixes **

File: Automation/NSG/deploy-nsg.ps1 [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+}
+
+# Error handling for general issues & nsg resource group deployment
+
+# Check if NSG already exists
+If (Get-AzNetworkSecurityGroup -Name $SecurityGroupName -ResourceGroupName $ResourceGroupName -ErrorAction Ignore) {
+    Write-Host -ForegroundColor Red "`nNetwork Security Group already exists, please try again with a different SecurityGroupName.`n"
+} Else {
+    # Deploy with template + parameters
+    If (New-AzResourceGroupDeployment -Name $DeploymentNames -ResourceGroupName $ResourceGroupName -TemplateFile $DefaultTemplateFilePath -TemplateParameterObject $templateParams -Verbose) {
```
Comment (unicornsleuth) [inline Automation/NSG/deploy-nsg.ps1:175]: $DeploymentName, not $DeploymentNames - this caused a failure for me

**Review by unicornsleuth — CHANGES_REQUESTED** [review]
2 tiny changes - only one of them causing a failure : )

File: Automation/NSG/deploy-nsg.ps1 (hunk 4 above)
Comment (jleeboyd) [reply Automation/NSG/deploy-nsg.ps1]: Good catch.

File: Automation/NSG/deploy-nsg.ps1 (hunk 5 above)
Comment (jleeboyd) [reply Automation/NSG/deploy-nsg.ps1]: Fixed!

**Comment by jleeboyd** [comment]
Added in changes mentioned by @unicornsleuth in https://github.com/unicornsleuth/ad440-winter2021-thursday-repo/pull/1

**Review by unicornsleuth — APPROVED** [review]
Everything works! Woo!

**Pull request merged**